File: python/71_ordered_fractions.py
from math import floor

def ordered_fractions():
    # using Stern-Brocot tree formula: given two neighboring fractions a/b and c/d,
    # the next fraction that appears between them in a Farey sequence is (a + c)/(b + d)
    # repeating this n times results in the mediant (a + c*n)/(b + d*n)
    
    a, b, c, d = 2, 5, 3, 7
    # while b + d < 1_000_000:
        # a, b = a + c, b + d
    # return a
    
    # 5 + 7*n <= 1_000_000
    # 7*n <= 1_000_000 - 5
    # n <= 999_995/7
    
    return a + c * floor((1_000_000 - b) / d)
    
if __name__ == "__main__":
    print(ordered_fractions())

File: java/_71_OrderedFractions.java
import java.util.*;

public class _71_OrderedFractions{
    
    public static void main(String[] args){
        // A sequence of reduced proper fractions is called a Farey sequence.
        // Given two fractions a/b and c/d that are neighbors in a Farey sequence,
        // the next fraction to appear between them will be (a + c)/(b + d) in the Farey sequence of order (b + d).
        
        // Let F_n represent the Farey sequence of order n.
        // In F_8, 2/5 and 3/7 are neighboring fractions.
        // The next fraction to appear between them is (2 + 3)/(5 + 7) = 5/12, which first appears in F_12.
        // Repeating this process results in the following sequence for the left fraction:
        // 5/12, 8/19, 11/26, 14/33, 17/40, ...
        // which can be generalized to (2 + 3n)/(5 + 7n) for n = 1, 2, 3, ...
        
        // The greatest value of n for which 5 + 7n <= 1000000 is
        // n = floor((100000 - 5)/7) = floor(999995/7) = 142856
        // Therefore, the corresponding term in the sequence is
        // (2 + 3*142856)/(5 + 7*142856) = 428570/999997
        
        int a = 2, b = 5, c = 3, d = 7;
        int n = (int) Math.floor((1000000 - 5)/7);
        a += n*c;
        
        System.out.println("Numerator of fraction immediately left of 3/7 in the set of reduced proper fractions up to d <= 1000000: " + a); // 428570
    }
}